Login: Hide password field until user submits email or username

stephanethomas opened the pull request:
This pull request is another iteration after https://github.com/Automattic/wp-calypso/pull/19277 aiming at updating the [`Login` page](https://wordpress.com/log-in) and any related page to better support **passwordless accounts**.

Here the goal is simply to hide the password field when the `Login` page loads, and show it once the user has submitted the form with an email address or a username. In the future, we will check the type of authentication of the account via a new API endpoint when the user clicks the `Continue` button, and either display the password field for regular accounts, or send an email with a magic link for passwordless accounts. You can find the mockup of this new flow at p6qnuF-5T0-p2#comment-6882.

##### After Page Load

<img width="456" alt="screenshot" src="https://user-images.githubusercontent.com/594356/32700126-f0b3a1e8-c7c0-11e7-9d70-bcabbf6122cd.png">

##### After Submit

<img width="456" alt="screenshot" src="https://user-images.githubusercontent.com/594356/32700131-0105924a-c7c1-11e7-8c8f-75723a04d3e0.png">

##### When Linking Account

<img width="456" alt="screenshot" src="https://user-images.githubusercontent.com/594356/32701046-30e734b0-c7cf-11e7-89e9-11a623662c5c.png">

#### Testing instructions

You can check that the new Track events are correctly triggered by executing the following command in your browser's console, and then reloading the page:

```
localStorage.setItem( 'debug', 'calypso:analytics:tracks' );
```

1. Run `git checkout update/login-flow` and start your server, or open a [live branch](https://calypso.live/?branch=update/login-flow)

##### Default login

2. Open the [`Login` page](http://calypso.localhost:3000/log-in) in an incognito window
3. Assert that the `Password` field is hidden
4. Assert that the label of the submit button is `Continue`
5. Fill in the `Email Address or Username` field, and click this `Continue` button
6. Assert that the `Password` field is now displayed, and has focus
7. Assert that the label of the submit button is now `Log In`
8. Assert that you can still log in

##### Default login (linking flow)

2. Log in with a WordPress.com account that is already connected to Google
3. Navigate to the [`Social Login` page](http://calypso.localhost:3000/social-login)
4. Click the `Disconnect` button
5. Sign out from WordPress.com
6. Open the [`Login` page](http://calypso.localhost:3000/log-in)
7. Click the `Continue with Google` button
8. Authenticate with Google
9. Assert that a `We found a WordPress.com account with the email address [...]` message is shown
10. Assert that the `Password` field is displayed, and has focus
11. Assert that the label of the submit button is `Log In`
12. Assert that you can still log in

##### WooCommerce

2. Open the WooCommerce [`Home` page](https://woocommerce.com/) in an incognito window
3. Click the `SIGN IN WITH WORDPRESS.COM` link in the header
4. Replace https://wordpress.com with http://calypso.localhost:3000 in the address bar, and reload
5. Follow steps `#3` to `#8` of the default login

##### Other OAuth apps

2. Open the Akismet [`Home` page](https://akismet.com/) in an incognito window
3. Click the `Sign In` button in the header
4. Replace https://wordpress.com with http://calypso.localhost:3000 in the address bar, and reload
5. Follow steps `#3` to `#8` of the default login

#### Reviews

- [ ] Code
- [x] Product
- [x] Tests

yurynix commented on client/state/login/actions.js:
```diff
@@ -572,3 +575,39 @@ export const logoutUser = redirectTo => ( dispatch, getState ) => {
+	} );
+
+	if ( usernameOrEmail === '' ) {
+		const error = {
```
Why not an instance of `Error` ?

stephanethomas replied:
Here I'm just matching the data structure returned by `getErrorFromHTTPError()`:

https://github.com/Automattic/wp-calypso/blob/47a486525668a2637236c2329d9b8e4568a086bd/client/state/login/actions.js#L80-L84

stephanethomas commented on client/state/login/actions.js:
```diff
@@ -572,3 +575,39 @@ export const logoutUser = redirectTo => ( dispatch, getState ) => {
+			code: 'empty_username',
+			message: translate( 'Please enter a username or email address.' ),
+			field: 'usernameOrEmail',
+		};
```
This is a hardcoded error for now, the same one (and possibly some others) will be returned by our future API endpoint.

yurynix replied:
I don't think we need this action creator at all, it's enough to just dispatch the action..

see: https://github.com/Automattic/wp-calypso/pull/19688/commits/475bbe6033b6ea78f70f863cb5c226bd9ea5d5ee

stephanethomas commented on client/blocks/login/login-form.jsx:
```diff
@@ -109,20 +113,42 @@ export class LoginForm extends Component {
+			return;
+		}
+
+		this.props.recordTracksEvent( 'calypso_login_block_login_form_get_auth_type' );
```
I'm not entirely sure how all those new Tracks event should be named, neither if they make sense.

stephanethomas commented:
@alisterscott, @hoverduck: a heads-up that this pull request is going to break E2E tests:

```
Failed Test: [WPCOM] Editor: Posts (mobile) Basic Public Post @canary @parallel @jetpack @chrome 
 # attempts: 2
     output: 
06:56:51 Magellan child process start

06:56:58   [WPCOM] Editor: Posts (mobile)
06:56:58     Basic Public Post @canary @parallel @jetpack
06:56:58       ✓ Delete Cookies and Local Storage (41ms)
06:56:58       Publish a New Post
06:57:52         1) Can log in
06:57:52 FAILED: Taking screenshot of: 'https://calypso.live/log-in?branch=update/login-flow'
06:57:55   1 passing (undefinedms)
06:57:55   1 failing
06:57:55   1) [WPCOM] Editor: Posts (mobile) Basic Public Post @canary @parallel @jetpack Publish a New Post Can log in:
06:57:55      TimeoutError: Timed out waiting for element with css selector of '#user_pass, #password' to be settable to: '*********'
06:57:55 Wait timed out after 41291ms
06:57:55       at WebDriverError (node_modules/selenium-webdriver/lib/error.js:27:5)
06:57:55       at TimeoutError (node_modules/selenium-webdriver/lib/error.js:262:5)
06:57:55       at node_modules/selenium-webdriver/lib/promise.js:2201:17
06:57:55       at ManagedPromise.invokeCallback_ (node_modules/selenium-webdriver/lib/promise.js:1376:14)
06:57:55       at TaskQueue.execute_ (node_modules/selenium-webdriver/lib/promise.js:3084:14)
06:57:55       at TaskQueue.executeNext_ (node_modules/selenium-webdriver/lib/promise.js:3067:27)
06:57:55       at asyncRun (node_modules/selenium-webdriver/lib/promise.js:2927:27)
06:57:55       at node_modules/selenium-webdriver/lib/promise.js:668:7
06:57:55       at process._tickCallback (internal/process/next_tick.js:109:7)
06:57:55   From: Task: Timed out waiting for element with css selector of '#user_pass, #password' to be settable to: '*********'
06:57:55       at scheduleWait (node_modules/selenium-webdriver/lib/promise.js:2188:20)
06:57:55       at ControlFlow.wait (node_modules/selenium-webdriver/lib/promise.js:2517:12)
06:57:55       at thenableWebDriverProxy.wait (node_modules/selenium-webdriver/lib/webdriver.js:934:29)
06:57:55       at Object.setWhenSettable (lib/driver-helper.js:153:16)
06:57:55       at LoginPage.login (lib/pages/login-page.js:31:16)
06:57:55       at LoginFlow.login (lib/flows/login-flow.js:52:20)
06:57:55       at LoginFlow.loginAndStartNewPost (lib/flows/login-flow.js:61:8)
06:57:55       at Context.<anonymous> (specs/wp-post-editor-spec.js:318:27)
06:57:55       at runTest (node_modules/selenium-webdriver/testing/index.js:163:22)
06:57:55       at node_modules/selenium-webdriver/testing/index.js:184:16
06:57:55       at new ManagedPromise (node_modules/selenium-webdriver/lib/promise.js:1077:7)
06:57:55       at controlFlowExecute (node_modules/selenium-webdriver/testing/index.js:183:14)
06:57:55       at TaskQueue.execute_ (node_modules/selenium-webdriver/lib/promise.js:3084:14)
06:57:55       at TaskQueue.executeNext_ (node_modules/selenium-webdriver/lib/promise.js:3067:27)
06:57:55       at asyncRun (node_modules/selenium-webdriver/lib/promise.js:2974:25)
06:57:55       at node_modules/selenium-webdriver/lib/promise.js:668:7
06:57:55   From: Task: [WPCOM] Editor: Posts (mobile) Basic Public Post @canary @parallel @jetpack Publish a New Post Can log in
06:57:55       at Context.ret (node_modules/selenium-webdriver/testing/index.js:182:10)
```

We should update this test. Let me know how we should handle that when you have a few :).

hafizrahman commented:
For **Default Login**, 

> 6. Assert that the Password field is now displayed, and has focus

Does not happen for me. Instead none of the two fields are focused. This is tested on calypso.live using Chrome. Didn't see anything odd in the console. The rest of the steps are working correctly.

No other issues for **Default login (linking flow)**, **WooCommerce**, and **Other OAuth apps** so far.

Tug commented:
Why not implement the actual wpcom endpoint to get the account type along with this PR?
It's a good opportunity to test that they both work together imo.

I just don't understand why we're just adding code (and filling the holes with mocks) instead of actually shipping the feature.

Tug commented on client/blocks/login/login-form.jsx:
```diff
@@ -109,20 +113,41 @@ export class LoginForm extends Component {
+		const { usernameOrEmail } = this.state;
+		const { accountType } = this.props;
+
+		if ( isRegularAccount( accountType ) ) {
```
Could we check first if accountType is loaded, then split those 2 actions into 2 functions?
```js
onSubmitForm = event => {
    if ( ! this.props.accountTypeLoaded ) {
        return this.loadAccountType();
    }

    this.tryLogin();
};
```

stephanethomas replied:
Hell yes, that's a very good idea!

Tug commented on client/blocks/login/login-form.jsx:
```diff
@@ -231,28 +256,32 @@ export class LoginForm extends Component {
-							requestError.field === 'password' && (
-								<FormInputValidation isError text={ requestError.message } />
-							) }
+						{ ( linkingSocialUser || isRegularAccount( accountType ) ) && (
```
We could probably refactor this condition into a function
```js
isPasswordVisible() {
    return this.props.socialAccountIsLinking || isRegularAccount( this.props.accountType )
}
```

stephanethomas replied:
Makes sense indeed, I'll do that.

Tug commented on client/state/login/actions.js:
```diff
@@ -572,3 +575,43 @@ export const logoutUser = redirectTo => ( dispatch, getState ) => {
+				field: 'usernameOrEmail',
+			};
+
+			defer( () => {
```
Why do we need to reject to promise asynchronously here?

stephanethomas replied:
Because of an issue mentioned on Slack in p1510498319000092-slack-amber-dev, discussed in p1510504100000034-slack-amber-dev, and fixed by Yury in https://github.com/Automattic/wp-calypso/pull/19688/commits/b93bb3018169ed0a84f8062b7c8dc66141a779c2. Basically the `Email Address or Username` field is not focused the second time your click on the `Continue` button if this field is empty.

stephanethomas commented:
> Does not happen for me. Instead none of the two fields are focused.

Good catch, that's because of a regression introduced by https://github.com/Automattic/wp-calypso/commit/b93bb3018169ed0a84f8062b7c8dc66141a779c2. I'll fix that.

stephanethomas commented:
> Why not implement the actual wpcom endpoint to get the account type along with this PR?
It's a good opportunity to test that they both work together imo.

For the records, we've discussed this in Slack and decided to move forward with this pull request as it is.

stephanethomas commented:
Thanks for your feedback guys! I have addressed all comments, this is ready for another review.

alisterscott commented:
I tested it out and it seemed to work for me.

I do however have an issue with these types of forms in that it requires me to use password managers like 1Password twice and makes the logging in flow a lot more clunky.

Is there anything we can do to make it easier for password extension users (1Password, Lastpass etc.) easier to log in without having to use their password extension multiple times? This is also a problem on iOS Safari using password extensions.

For example, currently:

![current](https://user-images.githubusercontent.com/128826/32760832-d80574d8-c93b-11e7-8a85-3456d9a07396.gif)

With the PR:

![login with single field](https://user-images.githubusercontent.com/128826/32760983-989b2b3e-c93c-11e7-8663-2c4f1e1d5476.gif)

alisterscott commented:
Thanks for the heads-up WRT the e2e tests: I have updated the tests to use the Enter key between username and password and this means they now run on both implementations (password hiddent or not), so we can merge this and be confident the e2e tests will work correctly.

https://github.com/Automattic/wp-e2e-tests/pull/804

stephanethomas commented:
> I have updated the tests to use the Enter key between username and password and this means they now run on both implementations (password hiddent or not), so we can merge this and be confident the e2e tests will work correctly.

Thank you very much!

> Is there anything we can do to make it easier for password extension users (1Password, Lastpass etc.) easier to log in without having to use their password extension multiple times? This is also a problem on iOS Safari using password extensions.

Thanks for bringing that up Alister (the animated GIFs are super useful by the way). That's something we're aware of - we had a look to it, and found that this change indeed does make form filling more cumbersome for password managers that simulate keypresses (such as KeyPass). We may be able to keep the original behavior for password managers that use HTML to perform the filling though, but decided that wasn't our top priority at the moment.

I'm curious, how does you password manager work when logging into a [Google account](https://accounts.google.com/)? It's a similar flow, where only the email address field is displayed at first.

Tug commented:
@stephanethomas So, although I agreed at first that it should not be a priority, I didn't know this was going in production so early.

Thinking about it I'm not convinced that passwordless login is necessarily a better way to log in in all cases. It will help a fair amount of users who often forget their password or don't use a password manager, but this UX change will also make login a problem again for every users that had this problem solved already. Given the amount of users who use chrome and firefox and have seen the messages to have their password saved automatically, I don't think users who use password managers is a tiny subset anymore.

I have seen a very cool flow implemented on aliexpress.com which seamlessly integrates with chrome's password manager (you don't even have to validate login). I'll have a look at how they do that!

In the meantime could we have a discussion about this UX change. @cburton4 What do you think?

yurynix commented:
I agree with @Tug and @alisterscott that password managers are a UX issue we should address before moving forward with this.

cburton4 commented:
@stephanethomas I've attached the sketch for the solution that we discussed earlier below. Let me know if you change your mind on wanting something higher fidelity to work from.

![sketch](https://user-images.githubusercontent.com/9257264/32792785-46cb9cda-c932-11e7-8edc-ef30406de409.JPG)

alisterscott commented:
@stephanethomas:

> I'm curious, how does you password manager work when logging into a Google account? It's a similar flow, where only the email address field is displayed at first.

I tried this out, and whilst it takes the additonal click after entering the username, it has prefilled the password already, so maybe the password is a hidden field and it's prefilled by the password manager before 'showing' it?

Gif:

![google](https://user-images.githubusercontent.com/128826/32811234-5908a1a0-c9ea-11e7-8f95-979ec8f56bfe.gif)

yurynix commented:
@alisterscott 
It seems to be depended on _how_ you saved the password at the password manager:
https://support.1password.com/create-multi-page-login/

Tug commented:
@yurynix Interesting.. I don't think many 1password user set that up though.

It would be interesting to test the most common password managers once the new design is implemented.

yurynix commented:
What I'm saying is, many providers, Google and Amazon included doing a multi step login like the above, password managers should support that by now, so it shouldn't be a major blocker.

yurynix commented:
I've reviewed the code and I have no comments, if we agree on the UX we can merge this IMO.

yurynix commented on client/state/data-layer/wpcom/users/auth-options/index.js:
```diff
@@ -0,0 +1,66 @@
+);
+
+export default {
+	[ LOGIN_AUTH_ACCOUNT_TYPE_REQUESTING ]: [ getAuthAccountTypeRequest ],
```
the convention in the codebase seems to be to have a noun here, `LOGIN_AUTH_ACCOUNT_TYPE_REQUEST`

stephanethomas replied:
The problem in this pull request is that we need to trigger an event when the user submits the login form, and another one when we actually call the API. That's why I actually used both `LOGIN_AUTH_ACCOUNT_TYPE_REQUEST` and `LOGIN_AUTH_ACCOUNT_TYPE_REQUESTING` - a pattern that is used in [other parts of our code](https://github.com/Automattic/wp-calypso/blob/4924ba7c40685cb7370309e65df42b5f1a59ca60/client/state/action-types.js#L410-L411).

yurynix replied:
I'm sorry, I still don't really understand why we need that action, it's not used anywhere at the reducer, the loading state is derived from `LOGIN_AUTH_ACCOUNT_TYPE_REQUEST`.

Why we need it? there is no reducer for it, we can just use `LOGIN_AUTH_ACCOUNT_TYPE_REQUEST`
see: https://github.com/Automattic/wp-calypso/pull/19688/commits/475bbe6033b6ea78f70f863cb5c226bd9ea5d5ee

In the `MEDIA_REQUESTING` case, we can as well do the same, look at `MEDIA_REQUEST` and not `MEDIA_REQUESTING` `client/state/media/reducer` using `MEDIA_REQUESTING` instead using `MEDIA_REQUEST`.

yurynix commented on client/state/login/actions.js:
```diff
@@ -572,3 +576,51 @@ export const logoutUser = redirectTo => ( dispatch, getState ) => {
+ * @param {String} usernameOrEmail - id of the user
+ * @return {Function} a promise that will resolve once the authentication account type has been retrieved
+ */
+export const getAuthAccountType = usernameOrEmail => dispatch => {
```
It seems to me, since you've went with the data layer approach, the only thing should be here now is:
```js
export const getAuthAccountType = usernameOrEmail => ( {
	type: LOGIN_AUTH_ACCOUNT_TYPE_REQUESTING,
	usernameOrEmail,
} );
```

Also, since you went with the data layer approach, you can't get a promise from here, and therefor the tracking can't rely on the promise returned from here.

stephanethomas replied:
I'm not sure to understand the first part of your comment. Right now this function is responsible for checking whether the username is empty or not (this cannot be performed by the API as it would return a `404` since that information is part of the endpoint's url itself). It also triggers three distinct events.

The last part of your comment I totally agree with. I've struggled quite a bit with that, and I'll update the code as you suggested in another comment.

yurynix replied:
Why we're checking it here and not in the form?

yurynix commented on client/blocks/login/login-form.jsx:
```diff
@@ -97,25 +110,90 @@ export class LoginForm extends Component {
+		this.props
+			.getAuthAccountType( usernameOrEmail )
+			.then( () => {
+				this.props.recordTracksEvent( 'calypso_login_block_login_form_get_auth_type_success' );
```
Seems like tracking have to be off-loaded to the data layer, since we can't get promise from here.
[example](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom/sites/automated-transfer/initiate/index.js#L74)

stephanethomas replied:
Thanks for the pointer! I'll update the code accordingly.

yurynix commented:
I've checked the flow with 1Password, It offered to save the log in credentials as usual
and then filled them out ok.

I'm not sure how it'll work with already saved credentials.

IMO UX wise it looks good and we can move further

stephanethomas commented:
Thanks for having another look @yurynix. I've addressed your comment.

stephanethomas commented:
> I'm sorry, I still don't really understand why we need that action, it's not used anywhere at the reducer, the loading state is derived from LOGIN_AUTH_ACCOUNT_TYPE_REQUEST. Why we need it? there is no reducer for it, we can just use LOGIN_AUTH_ACCOUNT_TYPE_REQUEST. see: 475bbe6

@yurynix, did you see [my comment](https://github.com/Automattic/wp-calypso/pull/19688#discussion_r153274895)?

>  Right now this function is responsible for checking whether the username is empty or not (this cannot be performed by the API as it would return a 404 since that information is part of the endpoint's url itself).

This was to handle the case where the username is empty. With your change we display this nasty error:

![video](https://user-images.githubusercontent.com/594356/33314677-a99475ec-d42e-11e7-88d2-01946fe4f2f9.gif)

yurynix commented:
@stephanethomas sorry, I indeed missed it, so, why not check it in the form?
Sounds like a task for form validation...

stephanethomas commented:
We have to dispatch a Redux event in order to update the global state tree with an error, so it looked like it was the best place to do that. I don't mind moving that into the component if you find a nice/smart way to integrate that.

yurynix commented:
Lets let the browser handle that https://github.com/Automattic/wp-calypso/pull/19688/commits/096cce62eb8da69b0141b36f0423b16b92305a65:

Chrome:
![screen shot 2017-11-28 at 13 18 50](https://user-images.githubusercontent.com/326402/33317183-21e9db94-d43f-11e7-96fb-3ad02db632cb.png)

FireFox:
![screen shot 2017-11-28 at 13 18 42](https://user-images.githubusercontent.com/326402/33317211-3117ce1e-d43f-11e7-86c7-25843be1f7c7.png)

Safari:
![screen shot 2017-11-28 at 13 18 23](https://user-images.githubusercontent.com/326402/33317236-3f275920-d43f-11e7-80e3-2356dc4d0d15.png)

stephanethomas commented:
I'm not sure I agree with that change as it's not consistent with how other errors are displayed in this form, and in other parts of Calypso. I'll let @cburton4 decide.

cburton4 commented:
Field error handling should remain consistent with Calypso. See image below.

<img width="490" alt="screen shot 2017-11-28 at 10 49 44 am" src="https://user-images.githubusercontent.com/9257264/33329367-5fdbc436-d42a-11e7-93e7-12643b2687b7.png">

stephanethomas commented:
I've rebased, and reverted the last changes in order to restore the error's original look-and-feel.

Tug commented:
Product 👍

Tug commented:
Code looks good as well!

apeatling commented:
@stephanethomas Let's merge it before you have to rebase again. :)

stephanethomas commented:
I've rebased and fixed an error with local suggestions that would be triggered because of our switch to React 16. I've also fixed an issue with the password field which would not be reset when changing of username or email.

Tug commented:
LGTM 👍

cburton4 commented:
Is the message box for Akismet standard for how that message should display? The hierarchy seems off.
